### PR TITLE
Now that there is isPartial flag, InitFrom*() can be called with an incomplete info.

### PR DESCRIPTION
**Pull request by ziyan**
…ncomplete info.

cc @kanunikov-denys @kanbouchou 

```python
In [5]: env.GetKinBody('dc1_container').GetId()
Out[5]: '6'
```

## Before: The `id` of the body changed from `6` to `2`

```python
In [12]: env.LoadJSON(b, UpdateFromInfoMode.OnlySpecifiedBodiesExact)
2022-01-19 13:27:34,861 openrave [VERBOSE] [environment.cpp:169 EnvironmentBase::EnvironmentBaseInfo::DeserializeJSONWithMapping] body id='', isRobot=0
2022-01-19 13:27:34,862 openrave [VERBOSE] [environment.cpp:215 EnvironmentBase::EnvironmentBaseInfo::DeserializeJSONWithMapping] created new body id=''
2022-01-19 13:27:34,862 openrave [WARN] [jsonreader.cpp:270 JSONReader::ExtractAll] env=1, remove redundant entry 0 and replace with 1 with body name 'dc1_container'
2022-01-19 13:27:34,862 openrave [VERBOSE] [environment-core.h:2674 UpdateFromInfo] === UpdateFromInfo start, env=1 ===
2022-01-19 13:27:34,862 openrave [VERBOSE] [environment-core.h:2725 UpdateFromInfo] ==== body: env = 1, id = , name = dc1_container ===
2022-01-19 13:27:34,862 openrave [VERBOSE] [environment-core.h:2839 UpdateFromInfo] env=1, update existing body 6
2022-01-19 13:27:34,862 openrave [INFO] [kinbody.cpp:6123 UpdateFromInfoResult OpenRAVE::KinBody::UpdateFromKinBodyInfo] env=1, body 'dc1_container' do not update id '6' since update has empty id
2022-01-19 13:27:34,862 openrave [VERBOSE] [libopenrave.h:652 UpdateChildrenFromInfo] could not find existing pointer which matches and can update
2022-01-19 13:27:34,862 openrave [VERBOSE] [environment-core.h:2854 UpdateFromInfo] env=1, update body 6 from info result 3
2022-01-19 13:27:34,862 openrave [VERBOSE] [fclspace.h:583 FCLSpace::RemoveUserData] FCL User data removed from env 1 (userdatakey fclcollision0x5587f6464ae0) : dc1_container
2022-01-19 13:27:34,862 openrave [VERBOSE] [fclspace.h:123 FCLSpace::FCLKinBodyInfo::LinkInfo::Reset] env=1, resetting link dc1_container:base col=0x5587f7001210
2022-01-19 13:27:34,862 openrave [VERBOSE] [fclspace.h:583 FCLSpace::RemoveUserData] FCL User data removed from env 1 (userdatakey fclcollision0x5587f6543fe0) : dc1_container
2022-01-19 13:27:34,862 openrave [VERBOSE] [fclspace.h:123 FCLSpace::FCLKinBodyInfo::LinkInfo::Reset] env=1, resetting link dc1_container:base col=0x5587f74a3fb0
2022-01-19 13:27:34,862 openrave [VERBOSE] [environment-core.h:3872 _UnassignEnvironmentBodyIndex] env=1, removed body name="dc1_container" (environmentBodyIndex=8), recycle body index later
2022-01-19 13:27:34,862 openrave [INFO] [environment-core.h:3696 _EnsureUniqueId] env=1, tried renaming body from  -> 0 due to conflict, but conflict again. This is highly unlikely to happen.
2022-01-19 13:27:34,862 openrave [INFO] [environment-core.h:3696 _EnsureUniqueId] env=1, tried renaming body from  -> 1 due to conflict, but conflict again. This is highly unlikely to happen.
2022-01-19 13:27:34,862 openrave [DEBUG] [environment-core.h:3856 _AssignEnvironmentBodyIndex] env=1, recycled body envBodyIndex=8 for dc1_container. 0 remaining in pool
2022-01-19 13:27:34,862 openrave [VERBOSE] [kinbody.cpp:4982 KinBody::_ComputeInternalInformation] env=1, initialized dc1_container in 0.000023[s]
2022-01-19 13:27:34,862 openrave [VERBOSE] [fclspace.h:248 FCLSpace::FCLKinBodyInfoPtr fclrave::FCLSpace::InitKinBody] env=1, self=0, init body dc1_container (8)
2022-01-19 13:27:34,862 openrave [DEBUG] [jsonreader.cpp:306 JSONReader::ExtractAll] env=1, loaded 1 bodies in 622[us]
Out[12]: ([], [RaveGetEnvironment(1).GetKinBody('dc1_container')], [])

In [13]: RaveGetEnvironment(1).GetKinBody('dc1_container').GetId()
Out[13]: '2'
```

## After: The `id` of the body stayed at `6`

```python
In [6]: env.LoadJSON(b, UpdateFromInfoMode.OnlySpecifiedBodiesExact)
2022-01-19 14:10:44,421 openrave [VERBOSE] [environment.cpp:169 EnvironmentBase::EnvironmentBaseInfo::DeserializeJSONWithMapping] body id='', isRobot=0
2022-01-19 14:10:44,421 openrave [VERBOSE] [environment.cpp:215 EnvironmentBase::EnvironmentBaseInfo::DeserializeJSONWithMapping] created new body id=''
2022-01-19 14:10:44,422 openrave [WARN] [jsonreader.cpp:270 JSONReader::ExtractAll] env=1, remove redundant entry 0 and replace with 1 with body name 'dc1_container'
2022-01-19 14:10:44,422 openrave [VERBOSE] [environment-core.h:2674 UpdateFromInfo] === UpdateFromInfo start, env=1 ===
2022-01-19 14:10:44,422 openrave [VERBOSE] [environment-core.h:2725 UpdateFromInfo] ==== body: env = 1, id = , name = dc1_container ===
2022-01-19 14:10:44,422 openrave [VERBOSE] [environment-core.h:2839 UpdateFromInfo] env=1, update existing body 6
2022-01-19 14:10:44,422 openrave [INFO] [kinbody.cpp:6128 UpdateFromInfoResult OpenRAVE::KinBody::UpdateFromKinBodyInfo] env=1, body 'dc1_container' do not update id '6' since update has empty id
2022-01-19 14:10:44,422 openrave [VERBOSE] [libopenrave.h:604 UpdateChildrenFromInfo] could not find existing pointer which matches and can update
2022-01-19 14:10:44,422 openrave [VERBOSE] [environment-core.h:2854 UpdateFromInfo] env=1, update body 6 from info result 3
2022-01-19 14:10:44,422 openrave [VERBOSE] [fclspace.h:583 FCLSpace::RemoveUserData] FCL User data removed from env 1 (userdatakey fclcollision0x563edd7af3c0) : dc1_container
2022-01-19 14:10:44,422 openrave [VERBOSE] [fclspace.h:123 FCLSpace::FCLKinBodyInfo::LinkInfo::Reset] env=1, resetting link dc1_container:base col=0x563ede710a20
2022-01-19 14:10:44,422 openrave [VERBOSE] [fclspace.h:583 FCLSpace::RemoveUserData] FCL User data removed from env 1 (userdatakey fclcollision0x563ede798100) : dc1_container
2022-01-19 14:10:44,423 openrave [VERBOSE] [fclspace.h:123 FCLSpace::FCLKinBodyInfo::LinkInfo::Reset] env=1, resetting link dc1_container:base col=0x563edebccd10
2022-01-19 14:10:44,423 openrave [VERBOSE] [environment-core.h:3872 _UnassignEnvironmentBodyIndex] env=1, removed body name="dc1_container" (environmentBodyIndex=8), recycle body index later
2022-01-19 14:10:44,423 openrave [DEBUG] [environment-core.h:3856 _AssignEnvironmentBodyIndex] env=1, recycled body envBodyIndex=8 for dc1_container. 0 remaining in pool
2022-01-19 14:10:44,423 openrave [VERBOSE] [kinbody.cpp:4989 KinBody::_ComputeInternalInformation] env=1, initialized dc1_container in 0.000050[s]
2022-01-19 14:10:44,423 openrave [VERBOSE] [fclspace.h:248 FCLSpace::FCLKinBodyInfoPtr fclrave::FCLSpace::InitKinBody] env=1, self=0, init body dc1_container (8)

2022-01-19 14:10:44,423 openrave [DEBUG] [jsonreader.cpp:306 JSONReader::ExtractAll] env=1, loaded 1 bodies in 2512[us]
Out[6]: ([], [RaveGetEnvironment(1).GetKinBody('dc1_container')], [])

In [7]: env.GetKinBody('dc1_container').GetId()
Out[7]: '6'
```

[partialUpdate.json.txt](https://github.com/rdiankov/openrave/files/7894269/partialUpdate.json.txt)

### CODE REVIEW

**Comment by rdiankov** [comment]
@ziyan let's talk about this change, i have many questions.

**Comment by ziyan** [comment]
## With partial JSON trying to update properties:

- c: 4

## Originally designed flow:

1. existing `KinBody` has properties:

  - a: 1
  - b: 2
  - c: 3

2. `KinBody.ExtractInfo()` creates KinBodyInfo containing:

  - a: 1
  - b: 2
  - c: 3

3. `KinBodyInfo.DeserializeJSON()` takes in the partial JSON and becomes:

  - a: 1
  - b: 2
  - c: 4

4. `KinBody.UpdateFromInfo()` or `KinBody.InitFromInfo()` takes the **complete** `KinBodyInfo`:

  - a: 1
  - b: 2
  - c: 4

## Recent JSONReader changes introduced the flow:

1. a brand new `KinBodyInfo` created from just the partial JSON:

  - c: 4

2. `KinBody.UpdateFromInfo()` or `KinBody.InitFromInfo()` takes the **partial** `KinBodyInfo`:

  - c: 4

3. `KinBody.UpdateFromInfo()` works because it can handle **partial** `KinBodyInfo`, `KinBody.InitFromInfo()` fails because it is supposed to be initializing from **complete** `KinBodyInfo`

File: src/libopenrave/kinbody.cpp [hunk 1]
```diff
@@ -797,7 +797,10 @@ bool KinBody::InitFromKinBodyInfo(const KinBodyInfo& info)
         return false;
     }
 
-    _id = info._id;
+    if( !info._id.empty() ) {
```
Comment (ziyan) [inline src/libopenrave/kinbody.cpp:800]: @rdiankov Unfortunately, `info` here is `const`, so we cannot simply update its `id` before passing in.

**Comment by ziyan** [comment]
@rdiankov Passes `testopenrave` tests, still running gitlab pipeline.

**Comment by ziyan** [comment]
@rdiankov gitlab pipeline also passed.

**Comment by rdiankov** [comment]
finally~ thanks~
i removed your changes from the header

**Comment by ziyan** [comment]
Thanks!

**Pull request merged**